File: sdk/python/src/flare_redact/crypto.py
"""Keyed primitives shared by the hash, pseudonym and surrogate modes.

Everything here is HMAC-SHA256 over the caller's ``transform_secret``. There is
no key stretching and no salt: the point is a *deterministic* mapping, so the
same value produces the same output on every host and in every language, which
is what makes a pseudonymised dataset joinable. That also means the secret is
the only thing standing between the output and a dictionary attack on a small
domain, so it has to be a real secret — see ``spec/SPEC.md`` §8.
"""

from __future__ import annotations

import hashlib
import hmac

__all__ = ["hmac_sha256", "hmac_fingerprint", "derive_bytes"]


class MissingSecretError(ValueError):
    """Raised when a keyed transform is requested without a secret."""


def _key_bytes(secret: str) -> bytes:
    if not secret:
        raise MissingSecretError(
            "A non-empty transform_secret is required for deterministic protected transforms."
        )
    return secret.encode("utf-8")


def hmac_sha256(secret: str, message: str) -> bytes:
    """Raw HMAC-SHA256 digest of ``message`` under ``secret``."""
    return hmac.new(_key_bytes(secret), message.encode("utf-8"), hashlib.sha256).digest()


def hmac_fingerprint(secret: str, value: str, length: int = 16) -> str:
    """Lowercase hex of the first ``length`` bytes of HMAC-SHA256(secret, value)."""
    return hmac_sha256(secret, value)[:length].hex()


def derive_bytes(secret: str, context: str, length: int) -> bytes:
    """Expand ``secret`` into ``length`` bytes bound to ``context``.

    Counter-mode HMAC, matching ``deriveBytes`` in the reference implementation
    byte for byte: block *i* is ``HMAC(secret, context + "\\x00" + str(i))``.
    """
    if length <= 0:
        return b""
    out = bytearray()
    counter = 0
    while len(out) < length:
        out += hmac_sha256(secret, f"{context}\x00{counter}")
        counter += 1
    return bytes(out[:length])

File: sdk/go/crypto.go
package flareredact

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"strconv"
)

// ErrMissingSecret is returned when a keyed transform is requested without a
// TransformSecret. Falling back to unkeyed output would silently produce
// pseudonyms anyone could reproduce, so this is an error rather than a default.
var ErrMissingSecret = errors.New("flareredact: a non-empty TransformSecret is required for keyed transforms")

func hmacSHA256(secret, message string) []byte {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(message))
	return mac.Sum(nil)
}

// HMACFingerprint returns the lowercase hex of the first n bytes of
// HMAC-SHA256(secret, value). It is the basis of ModeHash.
func HMACFingerprint(secret, value string, n int) string {
	sum := hmacSHA256(secret, value)
	if n > len(sum) {
		n = len(sum)
	}
	return hex.EncodeToString(sum[:n])
}

// DeriveBytes expands secret into n bytes bound to context, using counter-mode
// HMAC. Block i is HMAC(secret, context + "\x00" + itoa(i)), which is what makes
// the derived stream identical in every flare-redact engine.
func DeriveBytes(secret, context string, n int) []byte {
	if n <= 0 {
		return nil
	}
	out := make([]byte, 0, n)
	for counter := 0; len(out) < n; counter++ {
		out = append(out, hmacSHA256(secret, context+"\x00"+strconv.Itoa(counter))...)
	}
	return out[:n]
}
